p2panda-rs/src/materializer/dag.rs

```diff
@@ -0,0 +1,300 @@
+
+        // If there are no edges then this is the end of this branch we should return None
+        if out_edges.len() == 0 {
+            None
+        } else {
+            Some(out_edges)
+        }
+    }
+
+    /// Find the initial starting node for this DAG (the node with no parent)
+    pub fn initial_root(&self) -> Option<Node> {
+        let mut root = None;
+        for (parent, child) in self.graph.iter() {
+            match parent {
+                Some(_) => continue,
+                None => root = Some(child.to_owned()),
+            }
+        }
+        root
+    }
+
+    /// Perform depth-first traversal of DAG, merging all forks, and returns an ordered list of
+    /// nodes.
+    pub fn topological(&mut self) -> Vec<Node> {
+        // Array of queued graph nodes
+        let mut queue: Vec<Node> = Vec::new();
+
+        // Topologically ordered graph nodes
+        let mut ordered_nodes: Vec<Node> = Vec::new();
+
+        // The root node of this graph
+        let root = self.initial_root();
+
+        // Insert root node into queue if it exists
+        match root {
+            Some(node) => queue.insert(0, node),
+            None => (),
+        }
+
+        // Pop next root node from end of queue.
+        // Continue while there are items in the queue.
+        while let Some(mut current_node) = queue.pop() {
+            // Push the current node into final array of ordered nodes. This means it has been
+            // visited, we don't need a visited nodes array as we are using a queue.
+            ordered_nodes.push(current_node.to_owned());
+
+            // Walk from this node until reaching a tip (leaf) of the graph (a node with no edges).
+            // edges are returned in alphabetical order which is how we consistently resolve concurrent edits
+            // (last write wins).
+            while let Some(mut out_edges) = self.node_out_edges(&mut current_node) {
+                // The next node we will visit
+                let next_edge = out_edges.remove(0);
+
+                // Any other target nodes are pushed to the queue for later walking
+                for edge in out_edges {
+                    queue.insert(0, edge.1.to_owned());
+                }
+
+                // Push the next node we are visiting to the ordered_nodes array
+                ordered_nodes.push(next_edge.1.clone());
+
+                // Set the new current_node
+                current_node = next_edge.1;
+            }
+        }
+        ordered_nodes
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::DAG;
+
+    use crate::entry::{sign_and_encode, Entry, LogId, SeqNum};
+    use crate::hash::Hash;
+    use crate::identity::KeyPair;
+    use crate::materializer::{marshall_entries, Edge};
+    use crate::message::{Message, MessageEncoded, MessageFields, MessageValue};
+
+    use rstest::{fixture, rstest};
+    use std::convert::TryFrom;
```

Comment:
External imports I would move to the top, then crate level imports and then module level.